src/systemctl/systemctl-list-units.c

```diff
@@ -105,11 +106,199 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        slice_units->num_units = num_units;
+
+        /* Insert the slice name and corresponding units into the hashmap */
+        r = hashmap_put(slice_unit_map, slice_name, slice_units);
+        if (r < 0) {
+                for (size_t i = 0; i < num_units; i++)
+                        mfree(slice_units->units[i]);
+                mfree(slice_units->units);
+                mfree(slice_units);
+                return log_oom();
+        }
+        return r;
+}
+
+/* Function to retrieve units under a slice from the hash map */
+static SliceUnits *get_units_under_slice(const char *slice_name) {
+        SliceUnits *slice_units;
+
+        assert(slice_name);
+        if (!slice_unit_map)
+                return NULL;
+
+        /* Retrieve units under slice */
+        slice_units = hashmap_get(slice_unit_map, slice_name);
+        if (!slice_units)
+                return NULL;
+
+        return slice_units;
+}
+
+static bool is_slice(const char *slice_name) {
+        assert(slice_name);
+
+        if (!slice_unit_map)
+                return false;
+
+        return hashmap_contains(slice_unit_map, slice_name);
+}
+
+static void free_slice_units(SliceUnits *slice_units) {
+        if (slice_units) {
+                for (size_t i = 0; i < slice_units->num_units; i++)
+                        mfree(slice_units->units[i]);
+                mfree(slice_units->units);
+                mfree(slice_units);
+        }
+}
+
+static void free_hash_map(void) {
+        SliceUnits *slice_units;
+        void *v, *k;
+
+      if (slice_unit_map) {
+        HASHMAP_FOREACH_KEY(v, k, slice_unit_map) {
+                slice_units = hashmap_get(slice_unit_map, k);
+                free_slice_units(slice_units);
+        }
+
+        hashmap_free(slice_unit_map);
+        }
+}
+
+/* Function to print units under a slice recursively */
+static int print_units_under_slice(const char *slice_name, int level) {
+        SliceUnits *slice_units = NULL;
+        char **units;
+        size_t num_units;
+
+        assert(slice_name);
+        assert(level >= 0);
+        /* Retrieve units under the current slice from the hashmap */
+        slice_units = get_units_under_slice(slice_name);
+        if (slice_units == NULL)
+                return log_oom();
+
+        for(int i = 0; i < level; ++i)
+                printf("|   ");
+        printf("|__ %s\n", slice_name);
+
+        /* Print the units under the current slice */
+        units = slice_units->units;
+        num_units = slice_units->num_units;
+        for (size_t i = 0; i < num_units; ++i) {
+
+                /* Print child slices recursively */
+                if (is_slice(units[i]))
+                        print_units_under_slice(units[i], level + 1);
+                else {
+                        for (int j = 0; j < level + 1; ++j)
+                                printf("|   ");
+                        printf("|___%s\n", units[i]);
+                }
+        }
+        return 0;
```

Comment:
instead of outputting something completely new, just rearrange the usual tabular output that output_units_list() generates.

For that, please add an additional columns to the table_new() call in output_units_list(): one that contains the slice the unit is in. This should be populated for each unit, but be hidden table_hide_column_from_display(). In --by-slice mode it should be used as the primary sort key however, see table_set_sort(). 

And then, simply change the unit name column so that it is prefixed with the tree branches as appropriate, if --by-slice mode is on.

Thus, we get the usual tabular output, the --by-slice option just adjusts the ordering to take slices into account, and draws the tree structure in front of the unit name

Reply:
Hello sir @poettering , I have been having some challenges implementing the table output this way, I have been able to add the slice row for each unit and also added the slice each unit belongs to in the slice column. However i have been have challenges as regards the sorting and how it positions the units on the table which prevents the branches from being prefixed appropriately. I tried sorting table_set_sort() just using the slice column and also using the slice column and unit column. But in both cases, the presence of many units which do not have any slice in the table does not allow the prefixing of the unit name with the tree branches to be properly done.
https://gist.github.com/devdekunle/b6ac92029a035bd91eea83f8e0edd8a2 
Above is the github gist to see the output of the sorting.
Please can i use another way where i can recursively add units to the table based on the slice name and also create a different table which do not include units without slices so that that table can be generated when the --by-slice mode is passed